### PR TITLE
Storage version garbage collector

### PR DESCRIPTION
**Pull request by roycaihw**
/kind feature

[KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md). Implement StorageVersion garbage collector, which removes stale storage version entries / objects for apiservers whose apiserver IDs don't exist anymore.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md
```

/sig api-machinery
/assign @caesarxuchao 
/hold 
wait for pending PRs to merge

### CODE REVIEW

File: test/integration/storageversion/gc_test.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+		if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(
+			context.TODO(), svName, metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("failed to cleanup valid storage version: %v", err)
+		}
+	})
+
+	t.Run("deleting an id should delete a storage version entry that it owns", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA, idB)
+		assertStorageVersionEntries(t, kubeclient, 2, idA)
+		deleteTestAPIServerIdentityLease(t, kubeclient, idA)
```
Comment (caesarxuchao) [inline test/integration/storageversion/gc_test.go:115]: Can you also make apiserver A and apiserver B to have different encoding version, and then verify that after idA is deleted, the commonEncodingVersion is set?

File: test/integration/storageversion/gc_test.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	// update the created sv with intended status
+	createdSV.Status = sv.Status
+	if _, err := client.InternalV1alpha1().StorageVersions().UpdateStatus(
+		context.TODO(), createdSV, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("failed to update store version status: %v", err)
+	}
+}
+
+func assertStorageVersionGC(t *testing.T, client kubernetes.Interface) {
```
Comment (caesarxuchao) [inline test/integration/storageversion/gc_test.go:155]: nit: maybe assertStorageVersionGone or assertStorageVersionDeleted is more self-explanatory?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 3]
```diff
@@ -69,6 +71,21 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
 			handler.ServeHTTP(w, req)
 			return
 		}
+		// Allow writes to namespace API. The system namespace is required for apiserver-identity
+		// lease to exist.
+		if requestInfo.APIGroup == "v1" && requestInfo.Resource == "namespaces" {
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:76]: For create requests, we cannot tell the name of the namespace from request info. 

I definitely don't like having this exception. Another option is to make sure the user and group are [APIServerUser and SystemPrivilegedGroup](https://github.com/kubernetes/kubernetes/blob/75463bda1d9103a46de127528ee7b3b96ceada54/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L69-L84). I printed the user info for such a request:

```
&{system:apiserver 77b648fd-ac93-4e2b-a5ba-421ac0ed3197 [system:masters] map[]}
```

[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:76]: One thing we should do is to limit to create requests only.

[the remark this answers is not in the document]
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:76]: A followup might be marking the [namespace client](https://github.com/kubernetes/kubernetes/blob/75463bda1d9103a46de127528ee7b3b96ceada54/pkg/controlplane/instance.go#L555-L556) in the bootstrap controller with special user agent, to further distinguish it from the rest loopback clients.

Currently how it looks 
- in local cluster: kube-apiserver/v1.20.0 (linux/amd64) kubernetes/47d7403
- in integration test: storageversion.test/v0.0.0 (linux/amd64) kubernetes/$Format

File: pkg/controller/storageversiongc/gc_controller.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+				continue
+			}
+			serverStorageVersions = append(serverStorageVersions, ssv)
+		}
+		if !hasStaleRecord {
+			continue
+		}
+		if len(serverStorageVersions) == 0 {
+			if err := c.kubeclientset.InternalV1alpha1().StorageVersions().Delete(context.TODO(), sv.Name, metav1.DeleteOptions{}); err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go]: good point. Done

**Comment by roycaihw** [comment]
The controller logic is ready for review. Suggestion on https://github.com/kubernetes/kubernetes/pull/95978#discussion_r521765181 is welcome. 

I do think we should put the controller as a leader-elected control loop in API server (written in the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md#garbage-collection)). Working on that

File: pkg/controller/storageversiongc/gc_controller.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		c.storageVersionQueue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("storage version %v failed with: %v", key, err))
+	c.storageVersionQueue.AddRateLimited(key)
+	return true
+}
+
+func (c *Controller) syncLease(name string) error {
```
Comment (lavalamp) [inline pkg/controller/storageversiongc/gc_controller.go:150]: I think `sync` is a confusing name since this only handles deletions.

File: pkg/registry/apiserverinternal/storageversion/strategy.go [hunk 6]
```diff
@@ -96,8 +103,71 @@ func (storageVersionStatusStrategy) PrepareForUpdate(ctx context.Context, obj, o
 
 	newSV.Spec = oldSV.Spec
 	metav1.ResetObjectMetaForStatus(&newSV.ObjectMeta, &oldSV.ObjectMeta)
+
+	status := apiserverinternal.ConditionFalse
```
Comment (lavalamp) [inline pkg/registry/apiserverinternal/storageversion/strategy.go:107]: I don't understand why we would set this condition in the server here. Can we split the condition into another PR?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 7]
```diff
@@ -69,6 +71,21 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
...
+			handler.ServeHTTP(w, req)
+			return
+		}
+		// Allow writes to the lease API in kube-system. The storage version API depends on the
+		// apiserver-identity leases to operate. Leases in kube-system are either apiserver-identity
+		// lease (which gets garbage collected when stale) or leader-election leases (which gets
+		// periodically updated by system components). Both types of leases won't be stale in etcd.
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:95]: We should probably just blanket-exclude all resources with known, transient lifetime. I guess that's just leases and events.

File: pkg/registry/apiserverinternal/storageversion/strategy.go [hunk 8]
```diff
@@ -96,8 +103,71 @@ func (storageVersionStatusStrategy) PrepareForUpdate(ctx context.Context, obj, o
...
+	}
+	condition := apiserverinternal.StorageVersionCondition{
+		Type:   apiserverinternal.AllEncodingVersionsEqual,
+		Status: status,
+		// Generation corresponds to the desired state, which is always
+		// 0 for storage versions, because StorageVersionSpec is empty.
+		ObservedGeneration: newSV.Generation,
+		Reason:             reason,
+		Message:            fmt.Sprintf("The common encoding version is: %s", commonEncodingVersion),
+	}
```
Comment (caesarxuchao) [inline pkg/registry/apiserverinternal/storageversion/strategy.go:123]: I think this condition should only be updated in three cases:
1. oldCommonEncodingVersion==nil, newCommonEncodingVersion!=nil,
2. oldCommonEncodingVersion!=nil, newCommonEncodingVersion==nil,
3. oldCommonEncodingVersion!=nil, newCommonEncodingVersion!=nil, and oldCommonEncodingVersion != newCommonEncodingVersion.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 9]
```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
 			handler.ServeHTTP(w, req)
 			return
 		}
+		// The system namespace is required for apiserver-identity lease to exist. Allow the apiserver
+		// itself to create namespaces.
+		// NOTE: with this exception, if the bootstrap client writes namespaces with a new version,
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:78]: I'm nervous that there may be more dependencies like this, this feature is behind a gate, right?

File: pkg/registry/apiserverinternal/storageversion/strategy.go [hunk 10]
```diff
@@ -96,8 +103,71 @@ func (storageVersionStatusStrategy) PrepareForUpdate(ctx context.Context, obj, o
...
+//    newCondition, LastTransitionTime is set to now if the new status differs from the old status)
+// 2. if a condition of the specified type does not exist (LastTransitionTime is set to now() if unset,
+// and newCondition is appended)
+func setStatusCondition(conditions *[]apiserverinternal.StorageVersionCondition, newCondition apiserverinternal.StorageVersionCondition) {
+	if conditions == nil {
+		return
+	}
+	existingCondition := findStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		if newCondition.LastTransitionTime.IsZero() {
```
Comment (caesarxuchao) [inline pkg/registry/apiserverinternal/storageversion/strategy.go:143]: Is this check necessary?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 11]
```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
...
+		// and the upgraded apiserver dies before updating the StorageVersion for namespaces, the
+		// storage migrator won't be able to tell these namespaces are stored in a different version in etcd.
+		// Because the bootstrap client only creates system namespace and doesn't update them. This can
+		// only happen if the upgraded apiserver is the first apiserver that kicks off namespace creation,
+		// or if a upgraded server that joins an existing cluster have new system namespaces (other
+		// than kube-system, kube-public, kube-node-lease) that need to be created.
+		u, hasUser := request.UserFrom(ctx)
+		if requestInfo.APIGroup == "" && requestInfo.Resource == "namespaces" &&
+			requestInfo.Verb == "create" && hasUser &&
+			u.GetName() == user.APIServerUser && contains(u.GetGroups(), user.SystemPrivilegedGroup) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:88]: Just checking the name is sufficient, I think.

I think before beta we need a way for a priviliged user to override this check, perhaps via query parameter or header, to recover a stuck cluster, should there be some dependency chain we didn't know about.

**Review by lavalamp — COMMENTED** [review]
I didn't look carefully at the tests, but here's my thoughts on the rest.

File: pkg/registry/apiserverinternal/storageversion/strategy.go (hunk 8 above)
Comment (caesarxuchao) [reply pkg/registry/apiserverinternal/storageversion/strategy.go:123]: I see that in setStatusCondition, you only set the lastTransition time if the condition.status has flipped, so it implemented the case 1 and 2 I mentioned above, but it doesn't include the case 3.

Instead of checking in setStatusCondition, can you do the check early on here?

File: pkg/controller/storageversiongc/gc_controller.go [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+	_, err := c.kubeclientset.CoordinationV1().Leases(metav1.NamespaceSystem).Get(context.TODO(), name, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		return err
+	}
+	storageVersionList, err := c.kubeclientset.InternalV1alpha1().StorageVersions().List(context.TODO(), metav1.ListOptions{})
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go:158]: correct. Will add more comments

File: pkg/registry/apiserverinternal/storageversion/strategy.go (hunk 6 above)
Comment (roycaihw) [reply pkg/registry/apiserverinternal/storageversion/strategy.go:107]: will do it in a separate PR

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 13]
```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
...
 		}
+		// The system namespace is required for apiserver-identity lease to exist. Allow the apiserver
+		// itself to create namespaces.
+		// NOTE: with this exception, if the bootstrap client writes namespaces with a new version,
+		// and the upgraded apiserver dies before updating the StorageVersion for namespaces, the
+		// storage migrator won't be able to tell these namespaces are stored in a different version in etcd.
+		// Because the bootstrap client only creates system namespace and doesn't update them. This can
+		// only happen if the upgraded apiserver is the first apiserver that kicks off namespace creation,
+		// or if a upgraded server that joins an existing cluster have new system namespaces (other
+		// than kube-system, kube-public, kube-node-lease) that need to be created.
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:84]: kube-system namespace is only created when the cluster is newly created. So the etcd corruption only happens if the cluster admin creates a cluster, the kube-system namespace is created, but the cluster creation fails, then the cluster admin creates a cluster of a different version which supports a different encoding version for namespaces, but keeps using the same etcd.

I think this is an extremely rare use case.

File: pkg/registry/apiserverinternal/storageversion/strategy.go [hunk 14]
```diff
@@ -96,8 +103,71 @@ func (storageVersionStatusStrategy) PrepareForUpdate(ctx context.Context, obj, o
...
+// findStatusCondition finds the conditionType in conditions.
+func findStatusCondition(conditions []apiserverinternal.StorageVersionCondition, conditionType apiserverinternal.StorageVersionConditionType) *apiserverinternal.StorageVersionCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+
+	return nil
+}
```
Comment (caesarxuchao) [inline pkg/registry/apiserverinternal/storageversion/strategy.go:173]: We need unit tests.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go (hunk 9 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:78]: two alpha feature gates are required: StorageVersionAPI and APIServerIdentity

I have tested locally, allowing `namespaces` and `leases` writes stops this feature from deadlock itself.

we can get signal of the filter breaking any bootstrap process from the alpha e2e test
/test pull-kubernetes-e2e-gce-alpha-features

File: test/integration/storageversion/gc_test.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+func createTestStorageVersion(t *testing.T, client kubernetes.Interface, ids ...string) {
+	sv := &apiserverinternalv1alpha1.StorageVersion{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: svName,
+		},
+	}
+	for i, id := range ids {
+		v := apiserverinternalv1alpha1.ServerStorageVersion{
+			APIServerID:       id,
+			EncodingVersion:   fmt.Sprintf("v%d", i),
```
Comment (caesarxuchao) [inline test/integration/storageversion/gc_test.go:128]: This is fragile. The encoding version depends on the order of the ids.

How about encodingVersion = fmt.Sprintf("v%s", strings.TrimPrefix(id, "id"))? Also validating the ids are prefixed with "id" at the beginning of this function, and otherwise t.Fatalf(...)

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go (hunk 11 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:88]: will add to beta requirement

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go (hunk 13 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:84]: I agree rolling back a just-created-and-crashed cluster without recreating etcd is weird.

The other case is possible. Adding a new kube-reserved namespace to an existing cluster could leave object with wrong encoding version in etcd.

File: pkg/registry/apiserverinternal/storageversion/strategy.go (hunk 14 above)
Comment (roycaihw) [reply pkg/registry/apiserverinternal/storageversion/strategy.go:173]: will move the strategy change to a separate PR

File: pkg/controller/storageversiongc/gc_controller.go [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+	defer klog.Infof("Shutting down storage version garbage collector")
+
+	klog.Infof("Starting storage version garbage collector")
+
+	if !cache.WaitForCacheSync(stopCh, c.leasesSynced, c.storageVersionSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	// Both lease deletion and storageversion update move slowly.
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go:93]: reworded

File: pkg/registry/apiserverinternal/storageversion/strategy.go (hunk 8 above)
Comment (roycaihw) [reply pkg/registry/apiserverinternal/storageversion/strategy.go:123]: moving to a separate PR

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go (hunk 13 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:84]: Ack.

Can we look into the body of the request to get the name of the namespace? We can explore in a followup.

File: pkg/controller/storageversiongc/gc_controller.go [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+		}
+		if !hasStaleRecord {
+			continue
+		}
+		if err := c.updateOrDeleteStorageVersion(&sv, serverStorageVersions); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return compileErrors(errors)
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go:182]: The Aggregate interface doesn't expose a way to convert `[]error` to `error`. People may have different opinions about how to format the aggregated error, so I don't think we want to put this into a utility package for now

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go (hunk 13 above)
Comment (roycaihw) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:84]: I think https://github.com/kubernetes/kubernetes/pull/95978#discussion_r522391599 is required for both:
- solving this leak
- providing way to recover a stuck cluster (not sure if the use case is bootstrap only, will think about it later)

I'm not sure if deserializing the body in this filter is achievable. Could be a faster workaround if it is not hard to do. Will explore

File: test/integration/storageversion/gc_test.go (hunk 15 above)
Comment (roycaihw) [reply test/integration/storageversion/gc_test.go]: that's better. Done

I'll rewrite it later to just have a `map[id string]v1alpha1.ServerStorageVersion` somewhere

File: pkg/controller/storageversiongc/gc_controller.go [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+	castObj := obj.(*apiserverinternalv1alpha1.StorageVersion)
+	c.enqueueStorageVersion(castObj)
+}
+
+func (c *Controller) onUpdateStorageVersion(oldObj, newObj interface{}) {
+	castNewObj := newObj.(*apiserverinternalv1alpha1.StorageVersion)
+	c.enqueueStorageVersion(castNewObj)
+}
+
+func (c *Controller) enqueueStorageVersion(obj *apiserverinternalv1alpha1.StorageVersion) {
```
Comment (caesarxuchao) [inline pkg/controller/storageversiongc/gc_controller.go:227]: Can you add a function comment saying this enqueues the storage version if it has entry for invalid apiserver?

File: pkg/controller/storageversiongc/gc_controller.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go:184]: My bad :( I thought `error` was private. I will update to use this util

**Review by caesarxuchao — COMMENTED** [review]
I have some minor comments, otherwise lgtm.

File: pkg/controller/storageversiongc/gc_controller.go [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+			lease.Labels[controlplane.IdentityLeaseComponentLabelKey] != controlplane.KubeAPIServer {
+			// we cannot find a corresponding identity lease in cache, enqueue the storageversion
+			c.storageVersionQueue.Add(obj.Name)
+			return
+		}
+	}
+}
+
+func (c *Controller) onDeleteLease(obj interface{}) {
+	castObj := obj.(*coordinationv1.Lease)
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go]: good catch. Fixed

**Comment by roycaihw** [comment]
Updated and squashed the commits. PTAL

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 20]
```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
 			handler.ServeHTTP(w, req)
 			return
 		}
+		// The system namespace is required for apiserver-identity lease to exist. Allow the apiserver
+		// itself to create namespaces.
+		// NOTE: with this exception, if the bootstrap client writes namespaces with a new version,
+		// and the upgraded apiserver dies before updating the StorageVersion for namespaces, the
+		// storage migrator won't be able to tell these namespaces are stored in a different version in etcd.
+		// Because the bootstrap client only creates system namespace and doesn't update them. This can
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:81]: nit: "...doesn't update them, this..."

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go [hunk 21]
```diff
@@ -69,6 +73,31 @@ func WithStorageVersionPrecondition(handler http.Handler, svm storageversion.Man
...
+		// only happen if the upgraded apiserver is the first apiserver that kicks off namespace creation,
+		// or if a upgraded server that joins an existing cluster have new system namespaces (other
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go:83]: nit:s/a/an, s/have/has

**Comment by caesarxuchao** [comment]
LGTM. @lavalamp can you take a pass? Thanks.

**Comment by roycaihw** [comment]
/test pull-kubernetes-e2e-gce-alpha-features

[previous runs](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/95978/pull-kubernetes-e2e-gce-alpha-features/1327030543635714048/) were still ongoing after 1h. Mostly I'm trying to make sure the cluster bootstrap works after https://github.com/kubernetes/kubernetes/pull/96444 got merged this afternoon

File: pkg/controller/storageversiongc/gc_controller.go [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+	castObj, ok := obj.(*coordinationv1.Lease)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		castObj, ok = tombstone.Obj.(*coordinationv1.Lease)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a Lease %#v", obj))
```
Comment (lavalamp) [inline pkg/controller/storageversiongc/gc_controller.go:251]: Should probably fix this in a followup to have cached the set of leases we care about, or to list them in this case. It should be rare.

Comment (roycaihw) [reply pkg/controller/storageversiongc/gc_controller.go:251]: I need to verify if such a case corrupts the lease informer's cache.

If not, the storage version worker will eventually garbage collect any storage version that has non-existing lease when resync kicks in (default 12 hours), regardless if we get the lease delete event or not. 

It could be a problem if leases that we don't care about are triggering too much tombstone, and causing the GC to spam log where it shouldn't
